### PR TITLE
Added SAS authentication for blob storage

### PR DESCRIPTION
**Pull request by devendra-rgb**
I modified the blob service client function to make authentication using sas token

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import internal from 'stream';
 type Config = {
     account: string;
     accountKey: string;
+    sastoken: string;
```
Comment (jakeFeldman) [inline src/index.ts:7]: Please camelCase. `sasToken`

File: src/index.ts [hunk 2]
```diff
@@ -30,16 +31,28 @@ function getServiceBaseUrl(config: Config) {
...
 function makeBlobServiceClient(config: Config) {
     const account = trimParam(config.account);
     const accountKey = trimParam(config.accountKey);
+    const sasToken=trimParam(config.sastoken)
+    
     const serviceBaseURL = getServiceBaseUrl(config);
+    //if accountKey doesnot contain value return below line
+    if(sasToken!=''){
+    const blobServiceClient = BlobServiceClient.fromConnectionString(`${serviceBaseURL}?${sasToken}`);
+    return blobServiceClient;
```
Comment (jakeFeldman) [inline src/index.ts:44]:
```suggestion
return BlobServiceClient.fromConnectionString(`${serviceBaseURL}?${sasToken}`);
```

File: src/index.ts [hunk 3]
```diff
@@ -30,16 +31,28 @@ function getServiceBaseUrl(config: Config) {
...
+    }
 
     const sharedKeyCredential = new StorageSharedKeyCredential(account, accountKey);
     const pipeline = newPipeline(sharedKeyCredential);
     return new BlobServiceClient(serviceBaseURL, pipeline);
 }
 
+
```
Comment (jakeFeldman) [inline src/index.ts:52]: Please remove this extra white space

**Review by jakeFeldman — CHANGES_REQUESTED** [review]
Thank you for the PR. This looks good. A few nits to clean up. Also, will you update the table in the readme to include this new var? It's in the usage section https://github.com/jakeFeldman/strapi-provider-upload-azure-storage#usage

**Comment by devendra-rgb** [comment]
Sure! I will work on these

File: README.md [hunk 4]
```diff
@@ -31,13 +31,15 @@ To enable the provider, create or edit the file at `./config/plugins.js`.
...
+
 module.exports = ({ env }) => ({
   upload: {
     config: {
       provider: "strapi-provider-upload-azure-storage",
       providerOptions: {
         account: env("STORAGE_ACCOUNT"),
-        accountKey: env("STORAGE_ACCOUNT_KEY"),
+        accountKey: env("STORAGE_ACCOUNT_KEY"),//either account key or sas token is enough to make authentication 
+        accountKey: env("STORAGE_ACCOUNT_SAS_TOKEN"),
```
[the remark this answers is not in the document]
Comment (devendra-rgb) [reply README.md:42]: Yeah I fixed it

**Comment by devendra-rgb** [comment]
@jakeFeldman Please check the new commit I also ran lint to check every thing is correct everything is fine

**Comment by jakeFeldman** [comment]
Changes look good. Let me run this locally and if everything looks good I'll get this added into the next version! Thanks again for the PR!

File: src/index.ts [hunk 5]
```diff
@@ -33,8 +34,12 @@ function getServiceBaseUrl(config: Config) {
 function makeBlobServiceClient(config: Config) {
     const account = trimParam(config.account);
     const accountKey = trimParam(config.accountKey);
+    const sasToken = trimParam(config.sasToken);
     const serviceBaseURL = getServiceBaseUrl(config);
-
+    //if accountKey doesn't contain value return below line
+    if (sasToken != '') {
+        return BlobServiceClient.fromConnectionString(`${serviceBaseURL}?${sasToken}`);
```
Comment (jakeFeldman) [inline src/index.ts:41]: When adding a sasToken I get an error here

the sasToken has a format of `?sv=....` 

Are you looking to use a connection string vs a sasToken?

Comment (devendra-rgb) [reply src/index.ts:41]: Yeah

**Comment by devendra-rgb** [comment]
please verify it now it works without a connection string

**Comment by jakeFeldman** [comment]
Changes look good. I'll test again tonight

**Comment by devendra-rgb** [comment]
Sure!

**Review by jakeFeldman — APPROVED** [review]

**Comment by devendra-rgb** [comment]
Thank you for the quick response waiting to see the release : )

**Comment by jakeFeldman** [comment]
@devendra-rgb I'm doing some further testing an I'mm unable to get the sas token to work. Did you have this working locally?

**Comment by devendra-rgb** [comment]
Yeah

**Comment by jakeFeldman** [comment]
Interesting I've added my sasToken and I get a 403 `[AzureStorage Blob Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature](https://stackoverflow.com/questions/24492790/azurestorage-blob-server-failed-to-authenticate-the-request-make-sure-the-value)`

**Comment by devendra-rgb** [comment]
I will look into it and get with a solution by tonight

**Comment by jakeFeldman** [comment]
Uploads are working but fetching the images are failing within the strapi ui

**Comment by devendra-rgb** [comment]
Oh Let me see this then I will have a look how it is forming the url for fetching

**Pull request merged**